dev/src/bulk-writer.ts

```diff
@@ -738,58 +738,62 @@ export class BulkWriter {
    * `flush()` or `close()` call.
    * @private
    */
-  private _sendCurrentBatch(flush = false): void {
+  private _scheduleCurrentBatch(flush = false): void {
     if (this._bulkCommitBatch._opCount === 0) return;
 
-    const tag = requestTag();
     const pendingBatch = this._bulkCommitBatch;
+    this._bulkCommitBatch = new BulkCommitBatch(this.firestore);
 
     // Use the write with the longest backoff duration when determining backoff.
     const highestBackoffDuration = pendingBatch.pendingOps.reduce((prev, cur) =>
       prev.backoffDuration > cur.backoffDuration ? prev : cur
     ).backoffDuration;
-    const backoffMsWithJitter = BulkWriter.applyJitter(highestBackoffDuration);
-    const delayMs = this._rateLimiter.getNextRequestDelayMs(
-      pendingBatch._opCount
+    const backoffMsWithJitter = BulkWriter._applyJitter(highestBackoffDuration);
+    const delayedExecution = new Deferred<void>();
+
+    if (backoffMsWithJitter > 0) {
+      delayExecution(() => delayedExecution.resolve(), backoffMsWithJitter);
+    } else {
+      delayedExecution.resolve();
+    }
+
+    delayedExecution.promise.then(() =>
+      this._sendCurrentBatch(pendingBatch, flush)
     );
-    const finalDelayMs = Math.max(backoffMsWithJitter, delayMs);
+  }
 
-    const delayedExecution = new Deferred<void>();
-    this._bulkCommitBatch = new BulkCommitBatch(this.firestore);
+  /**
+   * Sends the provided batch once the rate limiter does not require any delay.
+   */
+  private async _sendCurrentBatch(
+    batch: BulkCommitBatch,
+    flush = false
+  ): Promise<void> {
+    const tag = requestTag();
 
     // Send the batch if it is does not require any delay, or schedule another
     // attempt after the appropriate timeout.
-    if (finalDelayMs === 0) {
-      const underRateLimit = this._rateLimiter.tryMakeRequest(
-        pendingBatch._opCount
-      );
-      assert(
-        underRateLimit,
-        'RateLimiter should allow request if delayMs === 0'
-      );
-      delayedExecution.resolve();
+    const underRateLimit = this._rateLimiter.tryMakeRequest(batch._opCount);
+    if (underRateLimit) {
+      await batch.bulkCommit({requestTag: tag});
+      if (flush) this._scheduleCurrentBatch(flush);
```

Comment:
This codeflow still confused me, but I now know why. This sound like it is trying to re-send the same batch again. Should we call this `_scheduleNextBatch()`? FWIW, I think this function could also just be called `_sendBatch()` since it takes the batch as an argument now.

Reply:
Yeah, that makes more sense. Renamed.